queries.py

```diff
@@ -1,160 +1,161 @@
 import os
-import json
 import pandas as pd
 from databricks import sql
 
 SERVER_HOSTNAME = os.getenv("SERVER_HOSTNAME")
 HTTP_PATH = os.getenv("HTTP_PATH")
 ACCESS_TOKEN = os.getenv("ACCESS_TOKEN")
+PUBLIC_ONLY = os.getenv("PUBLIC_ONLY", "False").lower() in ("true", "1", "yes")
 
+class QueryService:
+
+    _instance = None
+
+    @staticmethod
+    def get_instance():
+        if QueryService._instance is None:
+            QueryService._instance = QueryService()
+        return QueryService._instance
+
+    def __init__(self):
+        self.country_whitelist = None
+
+    def execute_query(self, query):
+        """
+        Executes a query and returns the result as a pandas DataFrame.
+        """
+        conn = sql.connect(
+            server_hostname=SERVER_HOSTNAME,
+            http_path=HTTP_PATH,
+            access_token=ACCESS_TOKEN,
+        )
+        cursor = conn.cursor()
+        cursor.execute(query)
+        df = cursor.fetchall_arrow().to_pandas()
+        cursor.close()
+        conn.close()
+        return df
+
+    def _initialize_country_whitelist(self):
+        if PUBLIC_ONLY and self.country_whitelist is None:
+            query = """
+                SELECT country_name
+                FROM boost.data_availability
+                WHERE boost_public = 'Yes'
+            """
+            self.country_whitelist = self.execute_query(query)["country_name"].tolist()
+
+    def _apply_country_whitelist_filter(self, df):
+        if self.country_whitelist is not None and 'country_name' in df.columns:
+            return df[df['country_name'].isin(self.country_whitelist)]
+        return df
+
+    def fetch_data(self, query):
+        self._initialize_country_whitelist()
```

Comment:
why do we need to initialize the whitelist each time we fetch the data?
not enough to initialize it at the point of instance initialization?

Reply:
It's a lazy initialization – it initializes the whitelist the first time `fetch_data` was called, and not again afterwards (see the checks inside `_initialize_country_whitelist`). I chose not to initialize the whitelist in `__init__` because I don't want to initialize it during app start time – now that I think about it, it's probably unnecessary to defer this to query time as any changes to the public whitelist would still require a restart of the server. I'll update the code to do it in `__init__` as it makes the code more straightforward.